server/filestore.go

```diff
@@ -5365,16 +5380,38 @@ func (fs *fileStore) subjectDeleteMarkerIfNeeded(sm *StoreMsg, reason string) fu
 	seq, ts := fs.state.LastSeq+1, time.Now().UnixNano()
 	// Store it in the stream and then prepare the callbacks
 	// to return to the caller.
-	if err := fs.storeRawMsg(sm.subj, hdr, nil, seq, ts, ttl); err != nil {
+	if err := fs.storeRawMsg(subj, hdr, nil, seq, ts, ttl); err != nil {
 		return nil
 	}
 	cb, tcb := fs.scb, fs.sdmcb
 	return func() {
 		if cb != nil {
-			cb(1, int64(fileStoreMsgSize(sm.subj, hdr, nil)), seq, sm.subj)
+			cb(1, int64(fileStoreMsgSize(subj, hdr, nil)), seq, subj)
 		}
 		if tcb != nil {
-			tcb(seq, sm.subj)
+			tcb(seq, subj)
+		}
+	}
+}
+
+// Filestore lock must be held but message block locks must not be.
+// The caller should call the callback, if non-nil, after releasing
+// the filestore lock.
+func (fs *fileStore) subjectDeleteMarkersAfterPurge() func() {
+	if !fs.cfg.SubjectDeleteMarkers || fs.markers.Size() == 0 {
+		return nil
+	}
+	cbs := make([]func(), 0, fs.markers.Size())
```

Comment:
Why the cbs versus just doing inline? Locking things?

Reply:
Yes, so we aren't holding the FS locks while doing callbacks.